lib/logflare/alerting.ex

```diff
@@ -0,0 +1,242 @@
+  def upsert_alert_job(%AlertQuery{} = alert_query) do
+    Logflare.AlertsScheduler.put_job(%Citrine.Job{
+      id: alert_query.id,
+      schedule: alert_query.cron,
+      extended_syntax: false,
+      task: {:run_alert, [alert_query]}
+    })
+
+    {:ok, get_alert_job(alert_query)}
+  end
+
+  @doc """
+  Initializes and ensures that all alert jobs are created.
+  TODO: batching instead of loading whole table.
+  """
+  def init_alert_jobs do
+    AlertQuery
+    |> Repo.all()
+    |> Stream.each(fn alert_query ->
+      if get_alert_job(alert_query) == nil do
+        upsert_alert_job(alert_query)
+      end
+    end)
+    |> Stream.run()
+
+    :ok
+  end
+
+  @doc """
+  Performs the check lifecycle of an AlertQuery.
+
+  Send notifications if necessary configurations are set. If no results are returned from the query execution, no alert is sent.
+  """
+  @spec run_alert(AlertQuery.t()) :: :ok
+  def run_alert(%AlertQuery{} = alert_query) do
+    alert_query = alert_query |> Repo.preload([:user])
+
+    with {:ok, [_ | _] = results} <- execute_alert_query(alert_query) do
+      if alert_query.webhook_notification_url do
+        WebhookAdaptor.Client.send(alert_query.webhook_notification_url, %{
+          "result" => results
+        })
+      end
+
+      if alert_query.slack_hook_url do
+        SlackAdaptor.send_message(alert_query.slack_hook_url, results)
+      end
+
+      :ok
+    else
+      {:ok, []} ->
+        :ok
+
+      other ->
+        other
+    end
+  end
+
+  @doc """
+  Deletes an AlertQuery's Citrine.Job from the scheduler
+  noop if already deleted.
+
+  ### Examples
+    iex> delete_alert_job(%AlertQuery{})
+    :ok
+    iex> delete_alert_job(alert_query.id)
+    :ok
+  """
+  @spec delete_alert_job(AlertQuery.t() | number()) :: :ok
+  def delete_alert_job(%AlertQuery{id: id}), do: delete_alert_job(id)
+
+  def delete_alert_job(alert_id) do
+    Logflare.AlertsScheduler.delete_job(alert_id)
+  end
+
+  @doc """
+  Executes an AlertQuery and returns its results
+
+  Requires `:user` key to be preloaded.
+
+  ### Examples
+    iex> execute_alert_query(alert_query)
+    {:ok, [{"user_id" => "my-user-id"}]}
+  """
+  @spec execute_alert_query(AlertQuery.t()) :: {:ok, [map()]}
+  def execute_alert_query(%AlertQuery{user: %User{}} = alert_query) do
+    Logger.info("Executing AlertQuery | #{alert_query.name} | #{alert_query.id}")
+
+    with {:ok, transformed_query} <-
+           Logflare.Sql.transform(:bq_sql, alert_query.query, alert_query.user_id),
+         {:ok, %{rows: rows}} <-
+           Logflare.BqRepo.query_with_sql_and_params(
+             alert_query.user,
+             alert_query.user.bigquery_project_id || env_project_id(),
+             transformed_query,
+             [],
+             parameterMode: "NAMED",
+             maxResults: 1000,
+             location: alert_query.user.bigquery_dataset_location
+           ) do
+      {:ok, rows}
```

Comment:
Oh ... not catching the error here. What happens when it crashes with Citrine? We likely don't want it retrying a bunch out of the gate.

Catch it and log it? Actually let's send the crash as an alert! 

"Your query crashed, it will get ran again next time. If it continues to crash you might need to change your alert query."

Reply:
We can handle do the error handling in the error handling as a task.
I do like the alert notifying, definitely would be helpful for debugging.
Sending the alert message would be good as well, though it might lead to noise on some types of errors such as connection issues, which might result in all alerts spamming their channels.

Reply:
Since the max interval is 15 mins, i think it would be fine in such a case. It would be quite (or rather very) unlikely for every single alert to both execute at the same time and error out. 🤔

Reply:
TAsk https://www.notion.so/supabase/Logflare-Alerts-Error-handling-for-queries-with-errors-send-alert-to-user-061bb3f283bc405a8868e9505a34da9f?pvs=4